server/src/game/components/synopsis.rs

```diff
@@ -0,0 +1,135 @@
+use std::cmp::Ordering;
+
+use serde::Serialize;
+
+use crate::game::{game_conclusion::GameConclusion, phase::PhaseType, player::PlayerReference, role::Role, win_condition::WinCondition, Game};
+
+pub struct SynopsisTracker {
+    player_synopses: Vec<PartialPlayerSynopsis>
+}
+
+impl SynopsisTracker {
+    pub fn new(num_players: u8) -> Self {
+        SynopsisTracker {
+            player_synopses: (0..num_players).map(|_|
+                PartialPlayerSynopsis {
+                    crumbs: Vec::new()
+                }
+            ).collect(),
+        }
+    }
+
+    pub fn get(game: &Game, conclusion: GameConclusion) -> Synopsis {
+        Synopsis {
+            player_synopses: game.synopsis_tracker.player_synopses.iter()
+                .enumerate()
+                .map(|(player_index, player_synopsis)| 
+                    player_synopsis.get(
+                        PlayerReference::new_unchecked(player_index as u8)
+                            .get_won_game(game)
+                    )
+                ).collect(),
+            conclusion
+        }
+    }
+
+    pub fn on_role_switch(game: &mut Game, player: PlayerReference, _: Role, role: Role) {
+        let night = if matches!(game.current_phase().phase(), PhaseType::Night | PhaseType::Obituary) { 
+            Some(game.day_number())
+        } else {
+            None
+        };
+
+        let win_condition = player.win_condition(game).clone();
+
+        game.synopsis_tracker.add_crumb_to_player(player, night, role, win_condition);
+    }
+
+    pub fn on_convert(game: &mut Game, player: PlayerReference, _: WinCondition, win_condition: WinCondition) {
+        let night = if matches!(game.current_phase().phase(), PhaseType::Night | PhaseType::Obituary) { 
+            Some(game.day_number())
+        } else {
+            None
+        };
+
+        let role = player.role(game);
+        
+        game.synopsis_tracker.add_crumb_to_player(player, night, role, win_condition);
+    }
+
+    fn add_crumb_to_player(&mut self, player: PlayerReference, night: Option<u8>, role: Role, win_condition: WinCondition) {
+        if let Some(ref mut synopsis) = self.player_synopses.get_mut(player.index() as usize) {
+            synopsis.add_crumb(SynopsisCrumb { night, role, win_condition });
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Synopsis {
+    player_synopses: Vec<PlayerSynopsis>,
```

Comment:
VecMap<PlayerReference, V> is used almost everywhere else

I dont care though so you can keep it

Reply:
The reason it's not a map is because every player has a synopsis

Reply:
Yeah I know